git2-hyper/src/lib.rs

```diff
@@ -0,0 +1,270 @@
+/// information such as proxies, SSL information, etc.
+///
+/// This function is unsafe largely for the same reasons as
+/// `git2::transport::register`:
+///
+/// * The function needs to be synchronized against all other creations of
+///   transport (any API calls to libgit2).
+/// * The function will leak `handle` as once registered it is not currently
+///   possible to unregister the backend.
+///
+/// This function may be called concurrently, but only the first `handle` will
+/// be used. All others will be discarded.
+pub unsafe fn register(handle: hyper::Client<HttpsConnector<HttpConnector>>) {
+    static INIT: Once = Once::new();
+
+    let handle = Arc::new(Mutex::new(handle));
+    let handle2 = handle.clone();
+    INIT.call_once(move || {
+        git2::transport::register("http", move |remote| factory(remote, handle.clone())).unwrap();
+        git2::transport::register("https", move |remote| factory(remote, handle2.clone())).unwrap();
+    });
+}
+
+fn factory(
+    remote: &git2::Remote<'_>,
+    handle: Arc<Mutex<hyper::Client<HttpsConnector<HttpConnector>>>>,
+) -> Result<Transport, Error> {
+    Transport::smart(
+        remote,
+        true,
+        HyperTransport {
+            handle,
+            base_url: Arc::new(Mutex::new(String::new())),
+        },
+    )
+}
+
+impl SmartSubtransport for HyperTransport {
+    fn action(
+        &self,
+        url: &str,
+        action: Service,
+    ) -> Result<Box<dyn SmartSubtransportStream>, Error> {
+        let mut base_url = self.base_url.lock().unwrap();
+        if base_url.len() == 0 {
+            *base_url = url.to_string();
+        }
+        let (service, path, method) = match action {
+            Service::UploadPackLs => ("upload-pack", "/info/refs?service=git-upload-pack", "GET"),
+            Service::UploadPack => ("upload-pack", "/git-upload-pack", "POST"),
+            Service::ReceivePackLs => {
+                ("receive-pack", "/info/refs?service=git-receive-pack", "GET")
+            }
+            Service::ReceivePack => ("receive-pack", "/git-receive-pack", "POST"),
+        };
+        info!("action {} {}", service, path);
+        Ok(Box::new(HyperSubtransport {
+            handle: self.handle.clone(),
+            service,
+            url_path: path,
+            base_url: self.base_url.clone(),
+            method,
+            reader: None,
+            sent_request: false,
+        }))
+    }
+
+    fn close(&self) -> Result<(), Error> {
+        Ok(())
+    }
+}
+
+impl HyperSubtransport {
+    fn err<E: Into<Box<dyn error::Error + Send + Sync>>>(&self, err: E) -> io::Error {
+        io::Error::new(io::ErrorKind::Other, err)
+    }
+
+    fn execute(&mut self, data: &[u8]) -> io::Result<()> {
+        if self.sent_request {
+            return Err(self.err("already sent HTTP request"));
+        }
+
+        // FIXME: wrap a runtime here is definitely NOT a good idea
+        let rt = tokio::runtime::Runtime::new().unwrap();
```

Comment:
I think that you might want to take a [Handle](https://docs.rs/tokio/1.4.0/tokio/runtime/struct.Handle.html) for `register`, which you then clone and pass around.

Reply:
fixed in 165de0b